src/axom/primal/geometry/NURBSPatch.hpp

```diff
@@ -0,0 +1,2745 @@
+   * 
+   * \warning This method does NOT maintain the patch shape,
+   *  i.e. is not performing knot insertion/removal.
+   */
+  void setNumControlPoints_v(int npts)
+  {
+    SLIC_ASSERT(npts > getDegree_v());
+
+    m_controlPoints.resize(getNumControlPoints_u(), npts);
+
+    if(isRational())
+    {
+      m_weights.resize(getNumControlPoints_u(), npts);
+    }
+
+    m_knotvec_v.makeUniform(npts, getDegree_v());
+  }
+
+  /*!
+   * \brief Set the knot value in the u vector at a specific index
+   *
+   * \param [in] idx The index of the knot
+   * \param [in] knot The updated value of the knot
+   */
+  void setKnot_u(int idx, T knot) { m_knotvec_u[idx] = knot; }
+
+  /*!
+   * \brief Set the knot value in the v vector at a specific index
+   *
+   * \param [in] idx The index of the knot
+   * \param [in] knot The updated value of the knot
+   */
+  void setKnot_v(int idx, T knot) { m_knotvec_v[idx] = knot; }
+
+  /*! 
+   * \brief Set the u knot vector by an axom::Array
+   *
+   * \param [in] knots The new knot vector
+   */
+  void setKnots_u(const axom::Array<T>& knots, int degree)
+  {
+    m_knotvec_u = KnotVectorType(knots, degree);
+  }
+
+  /*! 
+   * \brief Set the v knot vector by an axom::Array
+   *
+   * \param [in] knots The new knot vector
+   */
+  void setKnots_v(const axom::Array<T>& knots, int degree)
+  {
+    m_knotvec_v = KnotVectorType(knots, degree);
+  }
+
+  /*! 
+   * \brief Set the u knot vector by a KnotVector object
+   *
+   * \param [in] knotVector The new knot vector
+   */
+  void setKnots_u(const KnotVectorType& knotVector)
+  {
+    m_knotvec_u = knotVector;
+  }
+
+  /*! 
+   * \brief Set the v knot vector by a KnotVector object
+   *
+   * \param [in] knotVector The new knot vector
+   */
+  void setKnots_v(const KnotVectorType& knotVector)
+  {
+    m_knotvec_v = knotVector;
+  }
+
+  /// \brief Returns the degree of the NURBS Patch on the first axis
+  int getDegree_u() const { return m_knotvec_u.getDegree(); }
+
+  /// \brief Returns the degree of the NURBS Patch on the second axis
+  int getDegree_v() const { return m_knotvec_v.getDegree(); }
+
+  /// \brief Returns the order of the NURBS Patch on the first axis
+  int getOrder_u() const { return m_knotvec_u.getDegree() + 1; }
+
+  /// \brief Returns the order of the NURBS Patch on the second axis
+  int getOrder_v() const { return m_knotvec_v.getDegree() + 1; }
+
+  /// \brief Return a copy of the knot vector on the first axis
+  KnotVectorType getKnots_u() const { return m_knotvec_u; }
+
+  /// \brief Return a copy of the knot vector on the first axis as an array
+  axom::Array<T> getKnotsArray_u() const { return m_knotvec_u.getArray(); }
+
+  /// \brief Return a copy of the knot vector on the second axis
+  KnotVectorType getKnots_v() const { return m_knotvec_v; }
+
+  /// \brief Return a copy of the knot vector on the second axis as an array
+  axom::Array<T> getKnotsArray_v() const { return m_knotvec_v.getArray(); }
```

Comment:
The distinction here might be a bit confusing. The former is the KnotVector instance, while the latter is the underlying array.

Reply:
Does this rework help clarify things?